### ISSUE TITLE
horovod hang with tensorflow_mnist.py, gdb show block on semaphonre in nsync call

### ISSUE THREAD

**Issue by xajxiang** [horovod/horovod#880]
Framework: (TensorFlow, Keras, PyTorch, MXNet) TensorFlow
Framework version: 1.13.1
Horovod version: 0.16.0
MPI version: openmpi4.0.0
CUDA version: 10.0
NCCL version: 2.3.4
Python version: 3.6.5
OS and version: ubuntu 16.04

**Your question:**
Please ask your question here.

I  want to run tensorflow_minist.py, but block on issues such as #877 #876 yesterday.
@alsrgv  suggest me to  try out tf-nightly-gpu.

1. 1 node with 4 v100 GPU card
2. mpirun -np 4 --allow-run-as-root -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH -x NCCL_P2P_DISABLE=1 -mca pml ob1 -mca btl ^openib python tensorflow_mnist.py

the demo seem to run correctly

3 .mpirun -np 4 --allow-run-as-root -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH -x NCCL_P2P_DISABLE=0 -mca pml ob1 -mca btl ^openib python tensorflow_mnist.py

the demo seem to hang

![1](https://user-images.githubusercontent.com/37759620/53772971-6147a000-3f23-11e9-9a81-0e453123e91b.png)

4. the nvidia-smi show below and the gdb show below

![2](https://user-images.githubusercontent.com/37759620/53772974-64db2700-3f23-11e9-807b-7cc26cf4e328.png)
![3](https://user-images.githubusercontent.com/37759620/53772977-67d61780-3f23-11e9-80c4-a8a65d73c5e7.png)

**Comment by xajxiang** [comment]
@alsrgv  disable_p2p show exchange parameters with direct memeory, enable p2p show exchange parameter with p2p, which one has the highest performance? thanks.

what's our reason for #880

**Comment by alsrgv** [comment]
@xajxiang, sorry for delayed response.  Can you try upgrading NCCL to 2.3.7 and running again?  P2P should have better performance compared to SHM (P2P disable=1).

**Comment by ppwwyyxx** [comment]
I recently run into a hang with similar stack trace when using horovod: 
```
#0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38                                                          
#1  0x00007f9e2a00be04 in nsync::nsync_mu_semaphore_p_with_deadline(nsync::nsync_semaphore_s_*, timespec) ()
   from /private/home/XXX/.local/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#2  0x00007f9e2a00b5d1 in nsync::nsync_sem_wait_with_cancel_(nsync::waiter*, timespec, nsync::nsync_note_s_*) ()
   from /private/home/XXX/.local/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so         
#3  0x00007f9e2a008b14 in nsync::nsync_cv_wait_with_deadline_generic(nsync::nsync_cv_s_*, void*, void (*)(void*), void (*)(void*), timespec, nsync::nsync_note_s_*) ()                          
   from /private/home/XXX/.local/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#4  0x00007f9e2a009035 in nsync::nsync_cv_wait_with_deadline(nsync::nsync_cv_s_*, nsync::nsync_mu_s_*, timespec, nsync::nsync_note_s_*) ()
   from /private/home/XXX/.local/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so         
#5  0x00007f9e259ba29b in tensorflow::DirectSession::WaitForNotification(tensorflow::DirectSession::RunState*, tensorflow::CancellationManager*, long long) ()
   from /private/home/XXX/.local/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so         
#6  0x00007f9e259c12d1 in tensorflow::DirectSession::RunInternal(long long, tensorflow::RunOptions const&, tensorflow::CallFrameInterface*, tensorflow::DirectSession::ExecutorsAndKeys*, tensorflow
::RunMetadata*) () from /private/home/XXX/.local/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#7  0x00007f9e259c990a in tensorflow::DirectSession::Run(tensorflow::RunOptions const&, std::vector<std::pair<std::string, tensorflow::Tensor>, std::allocator<std::pair<std::string, tensorflow::Te
nsor> > > const&, std::vector<std::string, std::allocator<std::string> > const&, std::vector<std::string, std::allocator<std::string> > const&, std::vector<tensorflow::Tensor, std::allocator<tenso
rflow::Tensor> >*, tensorflow::RunMetadata*) () from /private/home/XXX/.local/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#8  0x00007f9e23365b25 in tensorflow::SessionRef::Run(tensorflow::RunOptions const&, std::vector<std::pair<std::string, tensorflow::Tensor>, std::allocator<std::pair<std::string, tensorflow::Tenso
r> > > const&, std::vector<std::string, std::allocator<std::string> > const&, std::vector<std::string, std::allocator<std::string> > const&, std::vector<tensorflow::Tensor, std::allocator<tensorfl
ow::Tensor> >*, tensorflow::RunMetadata*) () from /private/home/XXX/.local/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#9  0x00007f9e2356e8ea in TF_Run_Helper(tensorflow::Session*, char const*, TF_Buffer const*, std::vector<std::pair<std::string, tensorflow::Tensor>, std::allocator<std::pair<std::string, tensorflo
w::Tensor> > > const&, std::vector<std::string, std::allocator<std::string> > const&, TF_Tensor**, std::vector<std::string, std::allocator<std::string> > const&, TF_Buffer*, TF_Status*) ()
   from /private/home/XXX/.local/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so         
#10 0x00007f9e2356f505 in TF_SessionRun () from /private/home/XXX/.local/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#11 0x00007f9e2336280e in tensorflow::TF_SessionRun_wrapper_helper(TF_Session*, char const*, TF_Buffer const*, std::vector<TF_Output, std::allocator<TF_Output> > const&, std::vector<_object*, std:
:allocator<_object*> > const&, std::vector<TF_Output, std::allocator<TF_Output> > const&, std::vector<TF_Operation*, std::allocator<TF_Operation*> > const&, TF_Buffer*, TF_Status*, std::vector<_ob
ject*, std::allocator<_object*> >*) () from /private/home/XXX/.local/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so                            
#12 0x00007f9e2336295a in tensorflow::TF_SessionRun_wrapper(TF_Session*, TF_Buffer const*, std::vector<TF_Output, std::allocator<TF_Output> > const&, std::vector<_object*, std::allocator<_object*>
 > const&, std::vector<TF_Output, std::allocator<TF_Output> > const&, std::vector<TF_Operation*, std::allocator<TF_Operation*> > const&, TF_Buffer*, TF_Status*, std::vector<_object*, std::allocato
r<_object*> >*) () from /private/home/XXX/.local/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#13 0x00007f9e2331222e in _wrap_TF_SessionRun_wrapper () from /private/home/XXX/.local/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#14 0x000055f2a8fbc311 in _PyCFunction_FastCallDict () at /tmp/build/80754af9/python_1540319457073/work/Objects/methodobject.c:234
#15 0x000055f2a9045c1c in call_function () at /tmp/build/80754af9/python_1540319457073/work/Python/ceval.c:4837           
#16 0x000055f2a906871a in _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1540319457073/work/Python/ceval.c:3335
#17 0x000055f2a903fd8b in _PyFunction_FastCall (globals=<optimized out>, nargs=6, args=<optimized out>, co=<optimized out>) at /tmp/build/80754af9/python_1540319457073/work/Python/ceval.c:4919
#18 fast_function () at /tmp/build/80754af9/python_1540319457073/work/Python/ceval.c:4954                                 
#19 0x000055f2a9045cf5 in call_function () at /tmp/build/80754af9/python_1540319457073/work/Python/ceval.c:4858           
#20 0x000055f2a906871a in _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1540319457073/work/Python/ceval.c:3335
#21 0x000055f2a9040e78 in _PyEval_EvalCodeWithName (qualname=0x0, name=<optimized out>, closure=0x7f9d6248b550, kwdefs=0x0, defcount=0, defs=0x0, kwstep=2, kwcount=<optimized out>, kwargs=0x0,
    kwnames=0x0, argcount=<optimized out>, args=0x7f9db7820450, locals=0x0, globals=<optimized out>, _co=0x7f9dfb7066f0) at /tmp/build/80754af9/python_1540319457073/work/Python/ceval.c:4166
#22 PyEval_EvalCodeEx () at /tmp/build/80754af9/python_1540319457073/work/Python/ceval.c:4187
#23 0x000055f2a90418f4 in function_call () at /tmp/build/80754af9/python_1540319457073/work/Objects/funcobject.c:604
#24 0x000055f2a8fbc1de in PyObject_Call () at /tmp/build/80754af9/python_1540319457073/work/Objects/abstract.c:2261
#25 0x000055f2a9069d9a in do_call_core (kwdict=0x0, callargs=0x7f9db94a5938, func=0x7f9d8d05d0d0) at /tmp/build/80754af9/python_1540319457073/work/Python/ceval.c:5106
#26 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1540319457073/work/Python/ceval.c:3404
#27 0x000055f2a903f124 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1540319457073/work/Python/ceval.c:4166
```
Running TF1.13.1, cuda 10, NCCL 2.4.2, horovod 0.16.2.
I'm not sure if it's related to horovod or not since I've seen similar reports from others not using horovod.

**Comment by alsrgv** [comment]
Per [this](https://github.com/horovod/horovod/issues/1078#issuecomment-493816557), the issue may be solved by migrating to NCCL 2.4.7.

**Comment by ppwwyyxx** [comment]
Thanks for the pointer! Will try
